pkg/pillar/types/zfs.go

```diff
@@ -12,6 +13,9 @@ const (
 	// ZVolDevicePrefix controlled by mdev
 	ZVolDevicePrefix = "/dev/zvol"
 
+	// ZFSMetricsHistoryMax - max number of stored metrics
+	ZFSMetricsHistoryMax = 60
+
```

Comment:
Why do we need this for zfs metrics? We do not have it for any other metrics (and 99.9% of our metrics are cumulative since boot hence we can calculate averges even if the controller is missing some metrics message.)

The reason this is important is since the metrics message has all of the metrics (for device, app instances, etc) and it is limited to about 32 to 64kbyte is size.

Reply:
> I added this to store the history of metrics in Eve's memory, now will store the last 60 values, given that we do data collection every second, it turns out that in Eve there will always be access to metrics for the last minute.

Why do you need to collect them every second in the first place?
Aren't all/most of them cumulative counters since boot?

We haven't had any customer ask us about more frequent data but we do have customers asking how they can reduce the amount of bandwidth used (which we can do by having zedagent report metrics a lot less frequently than once a minute.)

Reply:
> Why do you need to collect them every second in the first place?
> Aren't all/most of them cumulative counters since boot?

I have only one reason for this - the accuracy of the information displayed on the controller. The shorter the interval between the collection of information from the meters, the more real and accurate the metrics and performance of the storage system are displayed. And this makes it more likely to detect anomalies with hardware or software problems in the storage system.
Sending a slice of metrics with an interval of more than one minute has little useful meaning, except to report that the counters are increasing and the disk has not yet failed.

> We haven't had any customer ask us about more frequent data but we do have customers asking how they can reduce the amount of bandwidth used (which we can do by having zedagent report metrics a lot less frequently than once a minute.)

As for reducing the amount of bandwidth used, I think that here it will not be possible to sit on two chairs at the same time, observing the quality of the displayed metrics and the rarity of the frequency of sending them to the controller.
In this case, for clients who need to reduce the amount of bandwidth used, I can, depending on the value of the timer (by which we send metrics), do the following:
- Send metric slices in the full amount necessary for accurate calculations, if there is no need to reduce the amount of bandwidth used.
- If it is necessary to reduce the amount of bandwidth used, then it is possible to send a slice with a collection difference per minute or even less often (which is not of practical utility), or not to send this information for this configurations (rare messages with metric), since it is not critical information for our system.

Reply:
> Sending a slice of metrics with an interval of more than one minute has little useful meaning, except to report that the counters are increasing and the disk has not yet failed.

Plus it also allows the controller/UI to determine the rate of change to the counters (by comparing two sets whether those sets are 1 minute or 1 hour apart).
This is how we do all the other metrics.

I see no reason why these ZFS counters need any different treatment than the other metrics.

Reply:
Well, if that's by design then ok. I removed this value, and also simplified the code so as not to store anything, since we do not need to collect metrics with a difference per second.